app/src/main/java/ai/elimu/appstore/synchronization/AppListArrayAdapter.java

```diff
@@ -66,12 +94,179 @@ public View getView(int position, @Nullable View convertView, @NonNull ViewGroup
                 }
             });
         }
 
         return listItem;
     }
+
+
+    /**
+     * Downloads APK file and updates progress bar in the UI during the download.
+     */
+    private class DownloadApplicationAsyncTask extends AsyncTask<ApplicationVersionGson, Integer, Integer> {
+
+        private ApplicationVersionGson applicationVersionGson;
+
+        private ProgressBar progressBarDownloadProgress;
+        private TextView textViewDownloadProgress;
+
+        public DownloadApplicationAsyncTask(ProgressBar progressBarDownloadProgress, TextView textViewDownloadProgress) {
+            this.progressBarDownloadProgress = progressBarDownloadProgress;
+            this.textViewDownloadProgress = textViewDownloadProgress;
+        }
+
+        @Override
+        protected Integer doInBackground(ApplicationVersionGson... applicationVersionGsons) {
+            Timber.i("doInBackground");
+
+            applicationVersionGson = applicationVersionGsons[0];
+            Timber.i("applicationVersionGson.getFileUrl(): " + applicationVersionGson.getFileUrl());
+            Timber.i("applicationVersionGson.getFileSizeInKb(): " + applicationVersionGson.getFileSizeInKb());
+
+            // Reset to initial state
+            Integer fileSizeInKbsDownloaded = 0;
+            publishProgress(fileSizeInKbsDownloaded);
+
+            // Download APK file and store it on SD card
+            String fileUrl = BuildConfig.BASE_URL + applicationVersionGson.getFileUrl() +
+                    "?deviceId=" + DeviceInfoHelper.getDeviceId(context) +
+                    "&checksum=" + ChecksumHelper.getChecksum(context) +
+                    "&locale=" + UserPrefsHelper.getLocale(context) +
+                    "&deviceModel=" + DeviceInfoHelper.getDeviceModel(context) +
+                    "&osVersion=" + Build.VERSION.SDK_INT +
+                    "&applicationId=" + DeviceInfoHelper.getApplicationId(context) +
+                    "&appVersionCode=" + DeviceInfoHelper.getAppVersionCode(context);
+            Timber.i("fileUrl: " + fileUrl);
+
+            String fileName = applicationVersionGson.getApplication().getPackageName() + "-" + applicationVersionGson.getVersionCode() + ".apk";
+            Timber.i("fileName: " + fileName);
+
+            Timber.i("Downloading APK: " + applicationVersionGson.getApplication().getPackageName() + " (version " + applicationVersionGson.getVersionCode() + ", " + applicationVersionGson.getFileSizeInKb() + "kB)");
+
+//            File apkFile = ApkLoader.loadApk(fileUrl, fileName, context);
+            String urlValue = fileUrl;
+            // Copied from ApkLoader:
+            Timber.i("Downloading from " + urlValue + "...");
+
+            String language = Locale.getDefault().getLanguage();
+            File apkDirectory = new File(Environment.getExternalStorageDirectory() + "/.elimu-ai/appstore/apks/" + language);
```

Comment:
Do I understand this correctly? You take the local language of the device and store the apk in that folder. What if the device has a language we don't support yet?

Reply:
Whoops, that's wrong. Instead, it should fetch the the chosen locale set during the initial application launch which only contain supported languages:

![screenshot_20171001-202252](https://user-images.githubusercontent.com/1451036/31057636-671f64bc-a6e6-11e7-87a1-b4df1fd25ac9.png)

Thanks for letting me know, this could have caused some problems in the future. I'm fixing it now :-)